src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table-key-value/table-key-value.component.ts

```diff
@@ -57,41 +61,84 @@ export class TableKeyValueComponent implements OnInit, OnChanges {
   }
 
   useData() {
-    let temp = [];
     if (!this.data) {
       return; // Wait for data
-    } else if (_.isArray(this.data)) {
-      const first = this.data[0];
-      if (_.isPlainObject(first) && _.has(first, 'key') && _.has(first, 'value')) {
-        temp = [...this.data];
+    }
+    this.tableData = this.makePairs(this.data);
+  }
+
+  makePairs(data: any) {
+    let temp = [];
+    if (!data) {
+      return; // Wait for data
+    } else if (_.isArray(data)) {
+      temp = this.makePairsFromArray(data);
+    } else if (_.isPlainObject(data)) {
+      temp = this.makePairsFromObject(data);
+    } else {
+      throw new Error('Wrong data format');
+    }
+    return this.postProcessing(temp);
+  }
+
+  makePairsFromArray(data: any[]) {
+    let temp = [];
+    const first = data[0];
+    if (_.isPlainObject(first)) {
+      if (_.has(first, 'key') && _.has(first, 'value')) {
+        temp = [...data];
       } else {
-        if (_.isArray(first)) {
-          if (first.length === 2) {
-            temp = this.data.map(a => ({
-              key: a[0],
-              value: a[1]
-            }));
-          } else {
-            throw new Error('Wrong array format');
-          }
-        }
+        throw new Error('Wrong object array format: {key: string, value: any}[]');
       }
-    } else if (_.isPlainObject(this.data)) {
-      temp = Object.keys(this.data).map(k => ({
-        key: k,
-        value: this.data[k]
-      }));
     } else {
-      throw new Error('Wrong data format');
+      if (_.isArray(first)) {
+        if (first.length === 2) {
+          temp = data.map(a => ({
+            key: a[0],
+            value: a[1]
+          }));
+        } else {
+          throw new Error('Wrong array format: [string, any][]');
+        }
+      }
     }
-    this.tableData = temp.map(o => {
-      if (_.isArray(o.value)) {
-        o.value = o.value.join(', ');
-      } else if (_.isObject(o.value)) {
-        return;
+    return temp;
+  }
+
+  makePairsFromObject(data: object) {
+    return Object.keys(data).map(k => ({
+      key: k,
+      value: data[k]
+    }));
+  }
+
+  postProcessing(data: any[]) {
+    const temp = data.map((v) => this.convertValue(v)).filter(o => o); // Filters out undefined
+    if (!this.renderObjects) {
+      return temp;
+    }
+    temp.forEach((v, i) => {
+      if (_.isPlainObject(v.value)) {
+        temp.splice(i, 1);
+        this.makePairs(v.value).forEach(item => {
+          if (this.appendParentKey) {
+            item.key = v.key + ' ' + item.key;
+          }
+          temp.splice(i, 0, item);
+          i++;
+        });
       }
-      return o;
-    }).filter(o => o); // Filters out undefined
+    });
+    return temp;
+  }
+
+  convertValue(v: any) {
```

Comment:
Consider to make this method private.

Reply:
I will rename them with an underscore instead so that they are still testable.